### PR TITLE
Implement Pet store using react with hooks

### PR DESCRIPTION
**Pull request by HristoTrendafilov**

### CODE REVIEW

File: pet-store-react-hooks/src/infrastructure/components/Modal/Modal.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import React, { useEffect } from 'react';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/components/Modal/Modal.tsx]: Remove this file and it's entire folder.

File: pet-store-react-hooks/src/index.html [hunk 2]
```diff
@@ -4,5 +4,8 @@
     <meta charset="UTF-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1" />
   </head>
-  <body></body>
+  <body>
+    <div id="modal"></div>
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/index.html]: Remove this.

File: pet-store-react-hooks/src/Home/petsTable.css [hunk 3]
```diff
@@ -0,0 +1,93 @@
+table {
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/petsTable.css]: I want component css files to have the same name as the component that imports them - capitalize the first letter.

File: pet-store-react-hooks/src/infrastructure/api.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  try {
+    return apiResponse.json() as T;
+  } catch (err) {
+    reportError(err);
+    throw new Error(`Error parsing JSON response. ${apiErrorInfo}`);
+  }
+}
+
+export function getAllPetsAsync(): Promise<Pet[]> {
+  return fetchFromApiAsync<Pet[]>('/pet/all', 'GET');
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/api.ts]: This endpoint does not return a full Pet.

File: pet-store-react-hooks/src/infrastructure/api.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import type { Pet, PetKind } from '~infrastructure/api-types';
+
+import { reportError } from './utils';
+
+const apiBaseUrl = 'http://localhost:5150';
+const apiWaitTimeout = 5000;
+
+type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';
+async function fetchFromApiAsync<T>(
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/api.ts]: Rename this to `fetchJSON`.

File: pet-store-react-hooks/src/infrastructure/components/loadingIndicator/LoadingIndicator.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import './loadingIndicator.css';
+
+interface LoadingIndicatorProps {
+  delay?: number;
+}
+
+export function LoadingIndicator(props: LoadingIndicatorProps) {
+  const { delay } = props;
+
+  const [showIndicator, setShowIndicator] = useState<boolean>(false);
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/components/loadingIndicator/LoadingIndicator.tsx]: This feature is good, but I did not ask for it. We'll keep it for now.

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import { ErrorMessage } from '~infrastructure/components/errorMessage/ErrorMessage';
+import { LoadingIndicator } from '~infrastructure/components/loadingIndicator/LoadingIndicator';
+
+import { PetsTable } from './PetsTable';
+
+import './home.css';
+
+export function Home() {
+  const [loading, setLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string>('');
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: Use `whatever | undefined` for fetch related `useState`'s. Don't use default values like `[]` or empty strings.

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  const [allPets, setAllPets] = useState<Pet[]>([]);
+  const [petKindsMap, setPetKindsMap] = useState<Map<number, string>>(
+    new Map<number, string>()
+  );
+
+  const refreshPets = useCallback(async (fetchPetKinds: boolean = false) => {
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: Don't use this perameter. Check if the pet kinds are already fetched.

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    } catch (err) {
+      reportError(err);
+      setError('System error. Please contact the system administrator.');
+    } finally {
+      setLoading(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    document.title = 'Pet store';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: Why is this here. Just hardcode it in `index.html`

**Review by HristoKolev — COMMENTED** [review]
Fix and get back to me.

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 10]
```diff
@@ -61,7 +64,7 @@ export function Home() {
       <div className="all-pets-card-body">
         {loading && <LoadingIndicator />}
         {error && <ErrorMessage message={error} />}
-        {!error && !loading && (
+        {allPets && petKindsMap && !loading && (
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: Why `!loading` ?

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 11]
```diff
@@ -1,29 +1,32 @@
...
 import { getAllPetsAsync, getPetKindsAsync } from '~infrastructure/api';
-import type { Pet } from '~infrastructure/api-types';
+import type { PetListItem } from '~infrastructure/api-types';
 import { ErrorMessage } from '~infrastructure/components/errorMessage/ErrorMessage';
 import { LoadingIndicator } from '~infrastructure/components/loadingIndicator/LoadingIndicator';
 
 import { PetsTable } from './PetsTable';
 
-import './home.css';
+import './Home.css';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx:13]: Git and windows have failed you here. When you rename a file and the only change is in capitalization - git on windows does not count it as a change. As a result everything works fine on your machine but the rename of the files is not tracked on git and other people who pull that commit will have a broken build.

Search online how to fix it.

File: pet-store-react-hooks/src/infrastructure/api.ts [hunk 12]
```diff
@@ -1,12 +1,12 @@
-import type { Pet, PetKind } from '~infrastructure/api-types';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/api.ts]: Rename this file to `api-client.ts`.

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+import { ErrorMessage } from '~infrastructure/components/errorMessage/ErrorMessage';
+import { LoadingIndicator } from '~infrastructure/components/loadingIndicator/LoadingIndicator';
+
+import { PetsTable } from './PetsTable';
+
+import './Home.css';
+
+export function Home() {
+  const [loading, setLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string | undefined>(undefined);
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: You can skip passing `undefined` as a default value as it will be `undefined` anyway and TS knows that.

**Review by HristoKolev — COMMENTED** [review]
Fix and get back to me.

File: pet-store-react-hooks/src/Home/Home.tsx (hunk 10 above)
Comment (HristoTrendafilov) [reply pet-store-react-hooks/src/Home/Home.tsx]: Forgot about it while i was testing for the pet kinds

File: pet-store-react-hooks/src/Home/Home.tsx (hunk 11 above)
Comment (HristoTrendafilov) [reply pet-store-react-hooks/src/Home/Home.tsx:13]: Had to go to the ,git directory and under [core] had to change -> ignorecase = false

Comment (HristoTrendafilov) [reply pet-store-react-hooks/src/Home/Home.tsx:13]: Now for some reason in the repository the are two home folders - Home and home. One of which should be deleted

Comment (HristoKolev) [reply pet-store-react-hooks/src/Home/Home.tsx:13]: Move the styles to the correct directory and remove the incorrect one. Test your app after.

**Review by HristoKolev — COMMENTED** [review]
See comment about duplicate directory.

File: pet-store-react-hooks/src/Home/Home.tsx (hunk 11 above)
Comment (HristoTrendafilov) [reply pet-store-react-hooks/src/Home/Home.tsx:13]: Deleted the incorrect directory and tested everything.

File: pet-store-react-hooks/src/home/Home.tsx [hunk 14]
```diff
@@ -0,0 +1,73 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/home/Home.tsx]: Please rename `home` to `Home` to match the component file name. Use `git mv` to rename it so that you don't have any problems with git.

**Review by HristoKolev — COMMENTED** [review]
One last comment.

File: pet-store-react-hooks/src/home/Home.tsx (hunk 14 above)
Comment (HristoTrendafilov) [reply pet-store-react-hooks/src/home/Home.tsx]: Still had problems but i renamed it to Home2 and then back to Home

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        for (const kind of petKinds) {
+          map.set(kind.value, kind.displayName);
+        }
+
+        setPetKindsMap(map);
+        hasFetchedPetKinds.current = true;
+      }
+
+      const pets = await petsPromise;
+      pets.sort((x, y) => (x.petId > y.petId ? -1 : 1));
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: You know that you can do `pets.sort((x, y) => y.petId - x.petId);` right?

File: pet-store-react-hooks/src/infrastructure/components/errorMessage/ErrorMessage.tsx [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import type { CSSProperties } from 'react';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/components/errorMessage/ErrorMessage.tsx]: Rename the folders in `components`.

**Review by HristoKolev — COMMENTED** [review]
See comments,

File: pet-store-react-hooks/src/Home/Home.tsx (hunk 15 above)
Comment (HristoTrendafilov) [reply pet-store-react-hooks/src/Home/Home.tsx]: Yeah, its not like i search everytime how to sort things in javascript :D

**Review by HristoKolev — APPROVED** [review]
Good. Next - implement deleting a pet.

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx [hunk 17]
```diff
@@ -0,0 +1,86 @@
+import { useState } from 'react';
+
+import { deletePetAsync } from '~infrastructure/api-client';
+import type { PetListItem } from '~infrastructure/api-types';
+import { ErrorMessage } from '~infrastructure/components/ErrorMessage/ErrorMessage';
+import { Modal } from '~infrastructure/components/Modal/Modal';
+import { formatDate } from '~infrastructure/utils';
+
+import '~infrastructure/components/Modal/Modal.css';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx]: With webpack you only need to import a css file once,

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.css [hunk 18]
```diff
@@ -0,0 +1,23 @@
+.delete-pet-modal-wrapper {
+  width: 30rem;
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.css]: Don't use `rem`'s for spacing use pixels. You can use `rem` for specifying font size.

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+import './DeletePetModal.css';
+
+type DeletePetModalProps = {
+  pet: PetListItem;
+  petKindsMap: Map<number, string>;
+  // Question: because im handling things differently on this callback from the parent component
+  // is it OK to use one callback with a parameter, or two - onClose() , onDelete()?
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx]: Separate callbacks and you call both on success. One to notify the parent that the delete succeeded and one to close the modal.

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    } catch (err) {
+      reportError(err);
+      setError('System error. Please contact the system administrator.');
+    } finally {
+      setIsDeleting(false);
+    }
+  }
+
+  return (
+    <Modal onBackdropClick={() => !isDeleting && onClose(false)}>
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx]: Use `useCallback` every time you can.

File: pet-store-react-hooks/src/Home/Home.tsx [hunk 21]
```diff
@@ -64,8 +64,17 @@ export function Home() {
       <div className="all-pets-card-body">
         {loading && <LoadingIndicator />}
         {error && <ErrorMessage message={error} />}
-        {allPets && petKindsMap && (
-          <PetsTable pets={allPets} petKindsMap={petKindsMap} />
+        {/* Had to add !loading so that the table is hidden and the spinner is shown on refreshPets */}
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/Home.tsx]: You should clear the data when making the request instead.

**Review by HristoKolev — COMMENTED** [review]
I wrote some of the things but we have to have a call for the others.

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx [hunk 22]
```diff
@@ -1,53 +1,57 @@
...
 import './DeletePetModal.css';
 
 type DeletePetModalProps = {
   pet: PetListItem;
   petKindsMap: Map<number, string>;
-  // Question: because im handling things differently on this callback from the parent component
-  // is it OK to use one callback with a parameter, or two - onClose() , onDelete()?
-  onClose: (hasDeleted: boolean) => void;
+  onClose: () => void;
+  onDelete: () => void;
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx]: Rename it to `onDeleted`

File: pet-store-react-hooks/src/Home/PetsTable.tsx [hunk 23]
```diff
@@ -1,22 +1,17 @@
...
+
+import { PetTableRow } from './PetTableRow';
 
 import './PetsTable.css';
 
 interface PetsTableProps {
   pets: PetListItem[];
   petKindsMap: Map<number, string>;
-  refreshPets: () => void;
+  onForDelete: (pet: PetListItem) => void;
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/PetsTable.tsx]: Rename to `onDelete`

File: pet-store-react-hooks/src/Home/PetTableRow.tsx [hunk 24]
```diff
@@ -0,0 +1,41 @@
+import { useCallback } from 'react';
+
+import type { PetListItem } from '~infrastructure/api-types';
+import { formatDate } from '~infrastructure/utils';
+
+interface PetTableRowProps {
+  pet: PetListItem;
+  petKindsMap: Map<number, string>;
+  onDeleteButtonClicked: (pet: PetListItem) => void;
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/PetTableRow.tsx]: Rename to `onDelete`

File: pet-store-react-hooks/src/infrastructure/components/Modal/Modal.tsx [hunk 25]
```diff
@@ -13,25 +11,34 @@ interface ModalProps {
 export function Modal(props: ModalProps) {
   const { children, onBackdropClick } = props;
 
-  useEffect(() => {
-    const handleBackdropClick = (event: MouseEvent) => {
-      if (modalRoot && modalRoot === event.target) {
+  const modalBackdropRef = useRef<HTMLDivElement>(null);
+
+  // Question: With this function and useEffect() there should be a simpler way for handling
+  const handleBackdropClick = useCallback(
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/components/Modal/Modal.tsx]: Why are you doing all this? Just add the `onClick` handler on the backdrop directly.

**Review by HristoKolev — CHANGES_REQUESTED** [review]
See comments.

File: pet-store-react-hooks/src/Home/PetsTable.tsx [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+            <th colSpan={2}>Actions</th>
+          </tr>
+        </thead>
+        <tbody>
+          {pets.length > 0 &&
+            pets.map((pet) => (
+              <PetTableRow
+                key={pet.petId}
+                pet={pet}
+                petKindsMap={petKindsMap}
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/Home/PetsTable.tsx]: You can pass a single pet kind here and not the whole map.

File: pet-store-react-hooks/src/infrastructure/api-client.ts [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  if (!apiResponse.ok) {
+    throw new Error(
+      `Received non successful status code: ${apiResponse.status}. ${apiErrorInfo}`
+    );
+  }
+
+  try {
+    return apiResponse.json() as T;
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/infrastructure/api-client.ts]: You are missing an await here. If the promise rejects - you wont be able to catch it.

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx [hunk 28]
```diff
@@ -0,0 +1,90 @@
+import { useCallback, useState } from 'react';
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx:1]: Loading indicator doesn't show up.

**Review by HristoKolev — CHANGES_REQUESTED** [review]
See comments.

**Review by HristoKolev — CHANGES_REQUESTED** [review]
This version of the project looks different than the vanillajs version. Please make it look the same. I shouldn't be able to tell how this is built by just looking at it.

**Review by HristoKolev — CHANGES_REQUESTED** [review]
There are more differences between the 2 versions. I've added them to the trello ticket.

**Review by HristoKolev — CHANGES_REQUESTED** [review]
Good. Proceed with Edit/Add pet functionality.

File: pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx [hunk 29]
```diff
@@ -10,13 +10,17 @@ import './DeletePetModal.css';
 
 type DeletePetModalProps = {
   pet: PetListItem;
-  petKindsMap: Map<number, string>;
+  petKind: string | undefined;
   onClose: () => void;
   onDeleted: () => void;
 };
 
+// Question: I have created this typeguard function to check what type is passed in our component
```
Comment (HristoKolev) [inline pet-store-react-hooks/src/DeletePetModal/DeletePetModal.tsx]: Remove this.

**Review by HristoKolev — CHANGES_REQUESTED** [review]
